pinot-core/src/test/java/org/apache/pinot/queries/WithOptionQueriesTest.java

```diff
@@ -0,0 +1,197 @@
+        QueryOptionKey.MIN_BROKER_GROUP_TRIM_SIZE)) {
+
+      options.clear();
+      for (String value : new String[]{"-10000000000", "-2147483648", "-1", "2147483648", "10000000000"}) {
+        options.put(setting, value);
+
+        Assert.assertThrows(setting + " must be a number between 0 and 2^31-1, got: " + value,
+            IllegalStateException.class, () -> {
+              getBrokerResponse("SELECT x, count(*) FROM " + RAW_TABLE_NAME + " GROUP BY x", options);
+            });
+      }
+    }
+  }
+
+  @Test
+  public void testOptionParsingSuccess() {
+    HashMap<String, String> options = new HashMap<>();
+    List<Object> groupRows = new ArrayList();
+    groupRows.add(new Object[]{0d, 40L}); //four times 10 records because segment gets multiplied under the hood
+
+    // int values
+    for (String setting : Arrays.asList(QueryOptionKey.NUM_GROUPS_LIMIT,
+        QueryOptionKey.MAX_INITIAL_RESULT_HOLDER_CAPACITY, QueryOptionKey.MULTI_STAGE_LEAF_LIMIT,
+        QueryOptionKey.GROUP_TRIM_THRESHOLD, QueryOptionKey.MAX_STREAMING_PENDING_BLOCKS,
+        QueryOptionKey.MAX_ROWS_IN_JOIN, QueryOptionKey.MAX_EXECUTION_THREADS,
+        QueryOptionKey.MIN_SEGMENT_GROUP_TRIM_SIZE, QueryOptionKey.MIN_SERVER_GROUP_TRIM_SIZE,
+        QueryOptionKey.MIN_BROKER_GROUP_TRIM_SIZE, QueryOptionKey.NUM_REPLICA_GROUPS_TO_QUERY)) {
+
+      options.clear();
+      for (String value : new String[]{"0", "1", "10000", "2147483647"}) {
+        if (QueryOptionKey.NUM_GROUPS_LIMIT == setting && "0".equals(value)) { // numGroupsLimit=0 returns empty result
+          continue;
+        }
+
+        options.put(setting, value);
+        List<Object[]> rows =
+            getBrokerResponse("SELECT mod(x,1), count(*) FROM " + RAW_TABLE_NAME + " GROUP BY mod(x,1)",
+                options).getResultTable().getRows();
+        assertEquals(rows, groupRows);
+      }
+    }
+
+    //long values
+    for (String setting : Arrays.asList(TIMEOUT_MS, MAX_SERVER_RESPONSE_SIZE_BYTES, MAX_QUERY_RESPONSE_SIZE_BYTES)) {
+      options.clear();
+      for (String value : new String[]{"1", "10000", "9223372036854775807"}) {
+        options.put(setting, value);
+        List<Object[]> rows = getBrokerResponse("SELECT * FROM " + RAW_TABLE_NAME, options).getResultTable().getRows();
+        assertEquals(rows, _allRecords);
+      }
+    }
+  }
+
+  private void assertEquals(List actual, List expected) {
+    if (actual == expected) {
+      return;
+    }
+
+    if (actual == null || expected == null || actual.size() != expected.size()) {
+      Assert.fail("Expected \n" + expected + "\n but got \n" + actual);
+    }
+
+    for (int i = 0; i < actual.size(); i++) {
+      Object act = actual.get(i);
+      Object exp = expected.get(i);
+
+      if (act instanceof Object[]) {
+        Assert.assertEquals(act, exp);
+      } else {
+        Assert.assertEquals(act, exp);
+      }
+    }
```

Comment:
```suggestion
      Assert.assertEquals(actual.get(i), expected.get(i));
```
nit: can be simplified